components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/DBBasedClaimMetadataManager.java

```diff
@@ -0,0 +1,211 @@
+    private final CacheBackedExternalClaimDAO externalClaimDAO = new CacheBackedExternalClaimDAO(new ExternalClaimDAO());
+
+    @Override
+    public List<ClaimDialect> getClaimDialects(int tenantId) throws ClaimMetadataException {
+
+        return this.claimDialectDAO.getClaimDialects(tenantId);
+    }
+
+    @Override
+    public ClaimDialect getClaimDialect(String claimDialectURI, int tenantId) throws ClaimMetadataException {
+
+        if (claimDialectURI == null || StringUtils.isBlank(claimDialectURI)) {
+            return null;
+        }
+
+        return this.claimDialectDAO.getClaimDialects(tenantId).stream()
+                .filter(claimDialect -> claimDialectURI.equals(claimDialect.getClaimDialectURI()))
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public void addClaimDialect(ClaimDialect claimDialect, int tenantId) throws ClaimMetadataException {
+
+        this.claimDialectDAO.addClaimDialect(claimDialect, tenantId);
+    }
+
+    @Override
+    public void removeClaimDialect(ClaimDialect claimDialect, int tenantId) throws ClaimMetadataException {
+
+        this.claimDialectDAO.removeClaimDialect(claimDialect, tenantId);
+        // When deleting a claim dialect the relevant external claim deletion is handled by the DB through
+        // ON DELETE CASCADE. Here we are removing the relevant cache entry.
+        externalClaimDAO.removeExternalClaimCache(claimDialect.getClaimDialectURI(), tenantId);
+    }
+
+    @Override
+    public List<LocalClaim> getLocalClaims(int tenantId) throws ClaimMetadataException {
+
+        return this.localClaimDAO.getLocalClaims(tenantId);
+    }
+
+    @Override
+    public LocalClaim getLocalClaim(String localClaimURI , int tenantId) throws ClaimMetadataException {
+
+        if (localClaimURI == null || StringUtils.isBlank(localClaimURI)) {
+            return null;
+        }
+
+        List<LocalClaim> localClaims = this.localClaimDAO.getLocalClaims(tenantId);
+        return localClaims.stream()
+                .filter(localClaim -> localClaimURI.equals(localClaim.getClaimURI()))
+                .findFirst().orElse(null);
+    }
+
+    @Override
+    public List<ExternalClaim> getExternalClaims(String externalClaimDialectURI, int tenantId)
+            throws ClaimMetadataException {
+
+        return this.externalClaimDAO.getExternalClaims(externalClaimDialectURI, tenantId);
+    }
+
+    @Override
+    public ExternalClaim getExternalClaim(String externalClaimDialectURI, String claimURI, int tenantId)
+            throws ClaimMetadataException {
+
+        if (claimURI == null || StringUtils.isBlank(claimURI)) {
+            return null;
+        }
+
+        return this.externalClaimDAO.getExternalClaims(externalClaimDialectURI, tenantId).stream()
+                .filter(externalClaim -> claimURI.equals(externalClaim.getClaimURI()))
+                .findFirst().orElse(null);
+    }
```

Comment:
check the previous comment

Reply:
Fixed using Optional<T> as the return type with 291bdd4dfd720dedee5ca8e2ab69363dfa4b08bc